pytket/binders/circuit/Circuit/add_op.cpp

```diff
@@ -517,11 +524,144 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
+             const py::kwargs &kwargs) {
+            // Get registers of circuit:
+            std::vector<QubitRegister> circ_qregs =
+                get_unit_registers<QubitRegister>(*circ);
+            std::vector<BitRegister> circ_cregs =
+                get_unit_registers<BitRegister>(*circ);
+
+            // Map name -> size for circuit registers:
+            std::map<std::string, std::size_t> circ_qreg_sizes;
+            for (const QubitRegister &qreg : circ_qregs) {
+              circ_qreg_sizes[qreg.name()] = qreg.size();
+            }
+            std::map<std::string, unsigned> circ_creg_sizes;
+            for (const BitRegister &creg : circ_cregs) {
+              circ_creg_sizes[creg.name()] = creg.size();
+            }
+
+            // Get box circuit:
+            std::shared_ptr<Circuit> box_circ = box.to_circuit();
+
+            // Get units of box:
+            const qubit_vector_t box_qubits = box_circ->all_qubits();
+            const bit_vector_t box_bits = box_circ->all_bits();
+
+            // Get registers of box:
+            std::vector<QubitRegister> box_qregs =
+                get_unit_registers<QubitRegister>(*box_circ);
+            std::vector<BitRegister> box_cregs =
+                get_unit_registers<BitRegister>(*box_circ);
+
+            // Map name -> size for box registers
+            std::map<std::string, std::size_t> box_qreg_sizes;
+            for (const QubitRegister &qreg : box_qregs) {
+              box_qreg_sizes[qreg.name()] = qreg.size();
+            }
+            std::map<std::string, unsigned> box_creg_sizes;
+            for (const BitRegister &creg : box_cregs) {
+              box_creg_sizes[creg.name()] = creg.size();
+            }
+
+            // Check that all units in the box are in a register:
+            for (const Qubit &qb : box_qubits) {
+              if (box_qreg_sizes.find(qb.reg_name()) == box_qreg_sizes.end()) {
+                throw CircuitInvalidity("Box contains non-register qubits.");
+              }
+            }
+            for (const Bit &cb : box_bits) {
+              if (box_creg_sizes.find(cb.reg_name()) == box_creg_sizes.end()) {
+                throw CircuitInvalidity("Box contains non-register bits.");
+              }
+            }
+
+            // Check that the sizes of the registers match up:
+            for (const auto &pair : box_qreg_sizes) {
+              if (circ_qreg_sizes.at(qregmap.at(pair.first)) != pair.second) {
+                throw CircuitInvalidity("Size mismatch in qubit registers");
+              }
+            }
+            for (const auto &pair : box_creg_sizes) {
+              if (circ_creg_sizes.at(cregmap.at(pair.first)) != pair.second) {
+                throw CircuitInvalidity("Size mismatch in bit registers");
+              }
+            }
+
+            // Populate the arguments (qubits then bits):
+            std::vector<UnitID> args;
+            for (const Qubit &qb : box_qubits) {
```

Comment:
It seems like for this to work, we need both `box_qubits = box_circ->all_qubits()` and `replacement.flatten_registers()` to handle `UnitID`s in lexicographic order. I'm wondering if we could explain this better, maybe with a comment.

Reply:
Done.